selfservice/strategy/oidc/provider_auth0.go

```diff
@@ -0,0 +1,194 @@
+}
+
+func (g *ProviderAuth0) oauth2() (*oauth2.Config, error) {
+	endpoint, err := url.Parse(g.config.IssuerURL)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	authUrl := *endpoint
+	tokenUrl := *endpoint
+
+	authUrl.Path = path.Join(authUrl.Path, "/authorize")
+	tokenUrl.Path = path.Join(tokenUrl.Path, "/oauth/token")
+
+	c := &oauth2.Config{
+		ClientID:     g.config.ClientID,
+		ClientSecret: g.config.ClientSecret,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  authUrl.String(),
+			TokenURL: tokenUrl.String(),
+		},
+		Scopes:      g.config.Scope,
+		RedirectURL: g.config.Redir(g.public),
+	}
+
+	return c, nil
+}
+
+func (g *ProviderAuth0) OAuth2(ctx context.Context) (*oauth2.Config, error) {
+	return g.oauth2()
+}
+
+func (g *ProviderAuth0) Claims(ctx context.Context, exchange *oauth2.Token) (*Claims, error) {
+	o, err := g.OAuth2(ctx)
+
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	client := o.Client(ctx, exchange)
+
+	u, err := url.Parse(g.config.IssuerURL)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+	u.Path = path.Join(u.Path, "/userinfo")
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	req.Header.Add("Authorization", "Bearer: "+exchange.AccessToken)
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+	defer resp.Body.Close()
+
+	// There is a bug in the response from Auth0. The updated_at field may be a string and not an int64.
+	// https://community.auth0.com/t/oidc-id-token-claim-updated-at-violates-oidc-specification-breaks-rp-implementations/24098
+	// We work around this by reading the json generically (as map[string]inteface{} and looking at the updated_at field
+	// if it exists. If it's the wrong type (string), we fill out the claims by hand.
+
+	// Once auth0 fixes this bug, all this workaround can be removed.
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	var values map[string]interface{}
+	if err := json.Unmarshal(b, &values); err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	if v, ok := values["updated_at"]; ok {
+		// There is an updated_at field. Look at the type. If float64 (default for json numbers, then
+		// we can just use the Claims. Otherwise we convert the value.
+		if _, ok := v.(string); ok {
+			// This is the bug. We need to convert.
+			c, err := auth0Claims(b)
+			if err != nil {
+				return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+			}
+
+			return c, nil
+		}
+	}
```

Comment:
I believe this could be simplified with tidwall/gjson along the lines of:


```suggestion
	updatedAt := 0
	// There is an updated_at field. Look at the type. If float64 (default for json numbers, then
	// we can just use the Claims. Otherwise we convert the value.
	if updatedAtField := gjson.GetBytes(b, "updated_at"); updatedAtField.Exists() && updatedAtField.IsString() {
			t, err := time.Parse(time.RFC3339, updatedAtField.String())
			updatedAt = t.Unix()
	}
	
	b, err := sjons.Delete(gjson.GetBytes(b, "updated_at")
	if err != nil {
		// ...
	}
```

This would remove the need for the `auth0Claims` function completely!

Reply:
Yes, this is good. I figured there was a better way to do this and I'm glad you found it. Those changes look great.